# MICRO582/EPI582 - Genomic Epidemiology Winter 2022

### Instructor(s): Evan Snitkin

### Description: 

### Date: Winter Term 2022

### Prerequisites:

## Overview:
------------

### Section - Working with sequence data at the command line
--------------------------------------------------------

#### Class 1 – Overview of genomic epidemiology
***




#### Class 2 – Introduction to Great Lakes and HPC
***




#### Class 3 – Sequence data formats
***




#### Class 4 – Illumina sequencing data and QC
***



### Section - Generating sequence data and performing annotation
------------------------------------------------------------

#### Class 5 – Genome assembly
***




#### Class 6 – Sequence databases
***




#### Class 7 – Rapid genome comparison
***




#### Class 8 – Genome annotation
***




#### Class 9 – Resistome analysis
***




#### Class 10 – Reference based variant calling
***



### Section - Basic phylogenetic analysis
-------------------------------------

#### Class 11 – Intro to RStudio
***




#### Class 12 – Basic phylogenetic analysis
***




#### Class 13 – Genome-wide association studies
***




#### Class 14 – Recombination detection
***



### Section - Hospital outbreak investigation
----------------------------------------------------

#### Class 15 – Overview of genomic epidemiology
***




#### Class 16 – Guest lecture on AMR in public health
***




#### Class 17 – Transmission in endemic settings
***




#### Class 18 – Regional outbreaks
***




#### Class 19 – Regional endemic spread
***



#### Class 20 – Nanopore sequencing
***



#### Class 21 – Guest lecture on AMR surveillance in the environment
***



### Section - Genomic epidemiology of SARS-CoV-2
--------------------------------------------

#### Class 22 – Guest lecture on COVID genomics in public health
***



#### Class 23 – SARS-CoV-2 genomics
***



#### Class 24 – Dated phylogenetic analysis
***


