Class 23 â€“ SARS-CoV-2 genomic data QC and cluster detection
===========================================================

Goals
----
- Learn about the impact that variant filtering can have on data analysis
- Explore the amount of genetic variation present at the early stages in the emergence of a exponentionally spreading pathogen
- Consider the importance of having contextual isolates in interpretting genomic clusters

Setup
-----
We are going to be working in RStudio again today. Take the following steps to get ready for the lab:

1. Start up your epid582 Rproject and create a new directory in it called class23 to hold data we will be analyzing today. 
2. Go on to Great Lakes and copy over the class 23 files to your working directory
3. Use cyberduck to bring the files down to the class23 directory you created on your own computer

Background on lab and data set
------------------------------
Today we will be analyzing some data generated by collaborators at Rush University Medical Center (RUMC) in Chicago. The objective of their study was to understand whether healthcare workers (HCWs) were acquiring SARS-CoV-2 patients. To this end they sequenced the following:

1. All SARS-CoV-2 positive samples from HCWs in March and April 2020
2. All SARS-CoV-2 positive samples from in-patients in March and April 2020
3. A background data of SARS-CoV-2 samples that were processed by the RUMC clinical microbiology lab during the same time period. This includes samples from a drive through testing site and emergency room visits.

We will go through the actual steps we took in analyzing this genomic data. These steps include the following:

1. Performing filtering of variants based on [recommendations](https://virological.org/t/issues-with-sars-cov-2-sequencing-data/473) at the time
2. Grouping individuals with identical sequences into putative transmission clusters
3. Examining the makeup of clusters (i.e. HCWs, patients and background)

Impact of variant filtering
---------------------------
With any sequencing data, taking proper care to filter out false positive variants is critical to the success of downstream analyses. With SARS-CoV-2 genomic data, it was recognized early on that there were certain error-prone regions and signatures of problem samples that required filtering strategies. Among the recomendations were:

1. Masking regions at either end of the chromosome (e.g. first/last ~50 bps), as these were highly error prone
2. Masking variants that showed evidence of homoplasy (i.e. emerged multiple times independently). It was suspected that these were either artefactual, highly mutagenic regions or under strong positive selection. Regardless of the reason, these variants would likely confound phylogenetic analyses.
3. Masking variants clustered on the genome, as these were suspected to be due to artefacts of sample preparation.
4. Removing of samples with large numbers of N's (i.e. uncalled/ambiguous bases), as these were unreliable.
5. Removing samples that were very distant from all other samples in the collection, as this was deemed to be more likely due to sample issues than a real phenomenon.

Based on these recomendations we implemented these filtering/masking strategies. Let's read in our masked and unmasked SARS-CoV-2 genomic alignments and see what impact our QC has on pairwise genetic distances among samples (i.e. genetic diversity in our collection).

```
library(ape)
library(ggplot2)
library(ggtree)

#Read in genomic and meta-data
genome_aln <- read.dna('class23/SARS_CoV_2_genomes.fa',
                       format = "fasta")

genome_aln_masked <- read.dna('class23/SARS_CoV_2_genomes_masked.fa',
                              format = "fasta")

meta_data <- read.table('class23/SARS_CoV_2_sample_data.txt',
                        sep = "\t")


#Compute pairwise distance among genomes
dist_mat <- dist.dna(genome_aln, 
                     as.matrix = T, 
                     model = "N", 
                     pairwise.deletion = T)
  
dist_mat_masked <- dist.dna(genome_aln_masked, 
                            as.matrix = T, 
                            model = "N", 
                            pairwise.deletion = T)


#Plot pairwise distance histograms
par(mfrow = c(2,1)) #Create two plots on one figure

hist(dist_mat[lower.tri(dist_mat)], 
     main = "Unmasked genomic distances",
     xlab = "Pairwise distance",
     xlim = c(0, max(dist_mat)),
     breaks = seq(0, max(dist_mat)+2,2)) #Plot histogram of lower triangle of matrix

hist(dist_mat_masked[lower.tri(dist_mat_masked)], 
     main = "Masked genomic distances",
     xlab = "Pairwise distance",
     xlim = c(0, max(dist_mat)),
     breaks = seq(0, max(dist_mat)+2,2)) #Plot histogram of lower triangle of matrix

```

Importance of having background genomes for genomic cluster detection
---------------------------------------------------------------------
Next, let's try and find putative transmission clusters including patients and HCWs. Note that the observation of these clusters isn't definitive evidence of transmission between them in the hospital, but helps us identify sets of individuals to perform a more detailed epidemiologic investigation on. As there was very little genetic variation among circulating strains this early in the pandemic it was common to use a genetic distance cutoff of 2 variants to define plausible recent transmission. Here we will be even more restrictive and require identical isolates, although we found similar results increasing the threshold to 2. 

First, let's look at the distribution of cluster sizes.

```
#Make cluster ID a character for the purposes of plotting
meta_data$ClusterID <- as.character(meta_data$ClusterID)

#Create subsetted meta-data with no background isolates
metadata_no_bkg <- subset(meta_data, Group != "Background")


#Plot size of clusters of patient/HCWs
cluster_size <- table(metadata_no_bkg$ClusterID)

#Plot the histogram
hist(cluster_size, 20)
```

Next, let's make a a barplot of cluster size, colored so we can see how it is composed of patients and healthcare workers.

```
#Subset data to only include clusters greater than size 1
meta_data_no_bkg_gt1 <- subset(metadata_no_bkg, ClusterID %in% names(cluster_size)[cluster_size > 1])


#Plot histogram of cluster size colored by patient/HCW
ggplot(data = meta_data_no_bkg_gt1, aes(x = ClusterID, fill = Group)) +
  geom_histogram(stat = "count")


#Plot prettier histogram
ggplot(data = meta_data_no_bkg_gt1, aes(x = reorder(ClusterID, ClusterID, function(x){length(x)}), fill = Group)) +
  geom_histogram(stat = "count") +
  scale_fill_manual(values=c("#999999", "darkgreen")) +
  labs(x = "Clusters with two or more individuals", y = "Number of individuals in cluster") +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

We observed several large clusters including both patients and healthcare workers, which is consistent with transmission between these two groups. Let's add our epidemiologically unrelated background isolates to see if this alters our view.

```
#Plot histogram of cluster size and include background isolates
meta_data_gt1 <- subset(meta_data, ClusterID %in% names(cluster_size)[cluster_size > 1])

ggplot(data = meta_data_gt1, aes(x = reorder(ClusterID, ClusterID, function(x){length(x)}), fill = Group)) +
  geom_bar(stat = "count") +
  scale_fill_manual(values=c("black", "#999999", "darkgreen")) +
  labs(x = "Clusters with two or more individuals", y = "Number of individuals in cluster") +
  theme(axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
```

Including the background isolates revealed that most of our clusters include genetically identical isolates from epidemiologically unrelated individuals, which makes our genomic clusters a bit less impressive. Even for clusters where no background isolates are included, we should still be skeptical based on this finding, as it indicates that at this stage of the pandemic it is not uncommon to observe genetically identical strains among individuals who may not have direct transmission linkages.

Ultimately, a more detailed epidemiologic investigation was performed that included:
  1. Using healthcare worker flow sheets that captures physical interactions between patients
  2. Considered the timing of individual isolates when assessing plausibility of transmission (i.e. physicial interaction when one individual was thought to be positive and the other negative)
  3. Considered plausible infectious windows when assessing transmission (i.e. putative source should be infectious based on reasonable estimates for shedding for asymptomatic individuals) 

The results of this investigation found minimal evidence of transmisison from patients to healthcare workers. One of the large clusters did end up having strong support, but was primarily thought to be due to transmission among healthcare workers.


Plotting clusters on the tree
-----------------------------
Finally, let's put our detected transmission clusters in phylogenetic context and see how we can overlay different types of meta-data on our tree.

```
##Plot tree with identified clusters highlighted
#Read in tree
tree <- read.tree('class23/SARS_CoV_2_genomes_masked.tree')

#Remove background isolates
tree_ptHCW <- drop.tip(tree,
                       row.names(meta_data)[meta_data$Group == "Background"])

#Create variables to highlight cluster and patient/hcw status
group_tip_ptHCW <- c(meta_data[tree_ptHCW$tip.label, 'Group'], 
                     rep(NA, Nnode(tree_ptHCW)))

cluster_tip_ptHCW <- c(meta_data[tree_ptHCW$tip.label, 'ClusterID'],
                       rep(NA, Nnode(tree_ptHCW)))

#Just highlight clusters > 1
clusters_1 <- names(cluster_size)[cluster_size == 1]

cluster_tip_ptHCW[cluster_tip_ptHCW %in% clusters_1] = NA;

#Plot tree!
ggtree(tree_ptHCW) + 
  geom_tippoint(aes(shape = group_tip_ptHCW, col = cluster_tip_ptHCW)) + 
  scale_color_discrete() +
  labs(shape = 'Group', col = "Cluster") 
```


